### ISSUE TITLE
Extract type information from a SimProcedure

### ISSUE THREAD

**Issue by axt** [angr/angr#412]
Is there a way to extract the type information from a `SimProcedure`?

What I'm thinking of: some of the `SimProcedure`s contains code like this:
```python
def run(self, sim_size):
        self.argument_types = {0: SimTypeLength(self.state.arch)}
        self.return_type = self.ty_ptr(SimTypeTop(sim_size))
```

I would like to access this info, but not sure how to do it. 
Also it would be nice if this would be accessable without running the SimProcedure.

**Comment by ltfish** [comment]
>  a way to extract the type information from a `SimProcedure`

This has been considered, but obviously no effort has been made so far.

**Comment by zardus** [comment]
An easy start would be to move the arguments dict out of `__init__` and make it a class attribute. We can always replace it with an instance-specific refined version at runtime (for example, if we hit something like printf, from which we can dynamically extract some type information).

**Comment by axt** [comment]
Is any internal tool using these? I didn't find any references to it in the code, except that they are being set, they are never read (only checked in default angr-dev repos).

Another minor problem I see with the current representation of arg types is, that it doesn't support variadic arguments (e.g: `printf(char*, ...)`). I promised a PR for this for `SimTypeFunction` in `s_type.py` to @rhelmot, and will send it sooner or later. Also this dict-like representation of `argument_types` 
```
    self.argument_types = {0: SimTypeLength(self.state.arch)}
```
is not capable of representing variadic arguments. (Not sure it is necessary at all, since if a `SimProcedure` has variadic arguments, then we need to calculate the types of it dynamically, as you mentioned, just asking it to have a completene representation of the signature).

BTW: Couldn't we use `SimTypeFunction` here to represent both `argument_types` and `return_types` of the `SimProcedure`?

Would you accept a PR for this, or you want to do it internally? I mean if I would do an initial patch for `malloc` and maybe `printf` (for dynamically refined types), could we start a discussion about it to come up with a proper design?

**Comment by ltfish** [comment]
> Is any internal tool using these? 

Not right now. We are in need of a design that allows us to integrate type information with binary analysis, but then we realize that the type information is useless if we don't care about type inference, data dependency, etc., which then boils down to having a data-flow analysis framework.

I'm lying down the foundation of a data-flow analysis framework right now, and as a first step towards using the type information in general, we would like to introduce the notion of "variables" into our binary analysis with an analysis routine called `VariableRecovery`. Then we will sit down and see how to better the design of type systems in angr.

As for now, it's rather a design issue than an implementation issue. While I'm happy to see PRs on it, I'd rather to have a discussion somewhere, no matter it is on GitHub, on Slack, or in person, to have a decent design first.

Here is a screenshot posted by me in angr core channel last night, which is the direct result of `VariableRecovery` - it's still alpha, but you can see the result here.

![variables in angr management](https://cloud.githubusercontent.com/assets/5071575/24766182/39a844f4-1aaf-11e7-84e3-ffbcd6e53da6.png)

**Comment by axt** [comment]
Unfortunatelly I didnt see this on the channel. 

I'm also ended up with the conclusion, that if I wanna do a proper DFA then I need to do variable recovery and a register analysis first, (maybe a points-to analysis too) and put the DFA framework on top of that. I asked this question, because on the long-term I want to propagate the type info back on the callgraph. But I don't need it at the moment, maybe I just looked to far into the future :-).

I did a lot of experiments with these, mostly built on top of the current DDG, so don't have any advanced solution just some plans. Need to run now, would be glad to discuss this later.
